### DISCUSSION TITLE
Is it possible to embed HTML or svg tags into the document?

### DISCUSSION THREAD

**Discussion by kengjungwa** [outline/outline#1652] [Self Hosting]
Hi! I'd would to ask if it's possible to embed HTML tag or svg tag into the document.

In the "/" menu, I don't see any menu item like "HTML code". Simply put the html tag as text also just render as simple text as shown in the image below.
![image](https://user-images.githubusercontent.com/1743535/99019242-7e54c300-258e-11eb-8481-f13da404d1f2.png)

If it's not currently support, is there any workaround?

Thanks!

**Comment by tommoor** [comment]
Hi, it is not possible due to the inherent security risks of allowing arbitrary user generated html. What, specifically, are you hoping to embed?

**Reply by kengjungwa** [reply]
@tommoor Thanks for your prompt reply! Basically, I'm trying to put a svg tag due to the fact that my team can easily continue editing on draw.io (as the source is in the tag itself) as well as zoom in as much as they can as the diagrams sometimes are huge and contains lots of text.

For example, the following svg image has been created on draw.io. 

Any advice on alternatives is also appreciated.

<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="346px" viewBox="-0.5 -0.5 346 326" content="&lt;mxfile host=&quot;app.diagrams.net&quot; modified=&quot;2020-11-13T05:02:34.996Z&quot; agent=&quot;5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36&quot; etag=&quot;TWsTMYMUKwOwfrqja4jg&quot; version=&quot;13.9.6&quot;&gt;&lt;diagram id=&quot;LzxX1RzwCH7sg5wKmse8&quot; name=&quot;Page-1&quot;&gt;7ZZNc5swEIZ/DcdkMDI2Pdau4xya9uBmOtObgmRQI1giFgP59V2MMMY0qWdq99STte8uQnr2wzhsmVRrw7P4AYTUjueKymGfHM+bTL0Z/TRK3Spz5rZCZJSwQb2wUa/Sil1YoYTMB4EIoFFlQzGENJUhDjRuDJTDsC3o4VszHsmRsAm5HqvflcC4VQPf7fV7qaK4e/PEtZ6Ed8FWyGMuoDyS2MphSwOA7SqpllI38Dou7XN3b3gPBzMyxXMeYDes/PHtZfuQP/ubardTL2Z+Y3fZcV3YC9vDYt0RoF0INhmLMlYoNxkPG09J+SYtxkSTNaHl+EDd7tKgrI4ke8C1hESiqSnEer3AwqoPOFu77NlPOqDxEffuOW7THR227onQwkL5PaDHeVDd3zFRiuq1rr9mXx7XZwGixGbNMqmipgdun7IkvdUAJ3igQK1SuTwUqkuiAeSoICUzcC/DcOr5A4bedMww+JcIvT8jNFCkQgoL5fp1xtwTRv6ZdTa7FiT/jEYUNJmsCQZjiCDletWriyHGPubzvhj38H5KxNqOWV4gnIk2h8KE8p3zMzubuYkkvhM3beOau7ybKCM1NcZuOIUvTp2NqO9R7ee6StvO9NwiRdVspYGL5hT7fTMDoggxd7yZpgsvngytIjxAPEocFSUOOedo4JlGgQZDSgppk72t0vpE4lpFzWwIKR+S9EVT4or+nT5aR6KE0G/N5mE5XGK0fBiOZ8/3x23D/HHb+Ndqm+n/BP5NAllwvQSS2X/c7H1Hn4hs9Qs=&lt;/diagram&gt;&lt;/mxfile&gt;" onclick="(function(svg){var src=window.event.target||window.event.srcElement;while (src!=null&amp;&amp;src.nodeName.toLowerCase()!='a'){src=src.parentNode;}if(src==null){if(svg.wnd!=null&amp;&amp;!svg.wnd.closed){svg.wnd.focus();}else{var r=function(evt){if(evt.data=='ready'&amp;&amp;evt.source==svg.wnd){svg.wnd.postMessage(decodeURIComponent(svg.getAttribute('content')),'*');window.removeEventListener('message',r);}};window.addEventListener('message',r);svg.wnd=window.open('https://viewer.diagrams.net/?client=1&amp;page=0&amp;edit=_blank');}}})(this);" style="cursor:pointer;max-width:100%;max-height:326px;"><defs/><g><ellipse cx="60" cy="40" rx="60" ry="40" fill="#ffffff" stroke="#000000" pointer-events="all"/><path d="M 164.56 200.6 C 150.45 189.36 146.47 169.01 155.25 152.93 C 164.04 136.85 182.88 129.99 199.39 136.86 C 215.9 143.73 224.99 162.22 220.7 180.16 C 216.4 198.1 200.02 209.99 182.35 208 M 164.56 182.1 L 164.56 200.6 L 145 195.05" fill="none" stroke="#000000" stroke-miterlimit="10" transform="rotate(80,185,170)" pointer-events="all"/><rect x="25" y="140" width="120" height="60" fill="#ffffff" stroke="#000000" pointer-events="all"/><path d="M 277.5 195 L 277.5 268.63" fill="none" stroke="#000000" stroke-miterlimit="10" pointer-events="stroke"/><path d="M 277.5 273.88 L 274 266.88 L 277.5 268.63 L 281 266.88 Z" fill="#000000" stroke="#000000" stroke-miterlimit="10" pointer-events="all"/><rect x="210" y="145" width="135" height="50" fill="none" stroke="none" pointer-events="all"/><g transform="translate(-0.5 -0.5)"><switch><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 133px; height: 1px; padding-top: 170px; margin-left: 211px;"><div style="box-sizing: border-box; font-size: 0; text-align: center; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: normal; word-wrap: normal; ">Loop pagination until load all products<br /></div></div></div></foreignObject><text x="278" y="174" fill="#000000" font-family="Helvetica" font-size="12px" text-anchor="middle">Loop pagination until...</text></switch></g><rect x="210" y="275" width="135" height="50" fill="none" stroke="none" pointer-events="all"/><g transform="translate(-0.5 -0.5)"><switch><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 133px; height: 1px; padding-top: 300px; margin-left: 211px;"><div style="box-sizing: border-box; font-size: 0; text-align: center; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: normal; word-wrap: normal; ">Loop pagination until load all products<br /></div></div></div></foreignObject><text x="278" y="304" fill="#000000" font-family="Helvetica" font-size="12px" text-anchor="middle">Loop pagination until...</text></switch></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://desk.draw.io/support/solutions/articles/16000042487" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>

**Reply by kengjungwa** [reply]
Attaching screenshot as github doesn't render the actual svg tag.
![image](https://user-images.githubusercontent.com/1743535/99031168-f92ad780-25a8-11eb-96e8-6b57ebcec34d.png)

**Reply by tommoor** [reply]
I think ideally we'd support embedding draw.io as a service, rather than SVG tags – this has been requested before, the way it would be achieved is through an iframe so it comes down to whether they allow iframing or embedding.

**Comment by ghost** [comment]
Notion supports arbitrary HTML embeds. Outline should too! 👍

**Reply by ghost** [reply]
I'm currently recommending we don't use Outline and use Notion instead because of this missing feature in Outline.

Security issues can be worked around (ignoring all scripts etc). Or worst case, this could be enabled with a per workspace flag.

**Comment by chenyf** [comment]
we also need support html tag, as markdown use it for alignment

**Comment by T2brozz** [comment]
html tags would be nice, because markdown can't everything.

**Comment by IvanLi-CN** [comment]
Provides the ability to insert html for users with professional capabilities, which may be a shortcut for adding rich text representations.

In my case, I want the `kbd` tag and its corresponding styles.

<kbd>⌘</kbd>

**Comment by T2brozz** [comment]
I would like to use it to extend outline. And custom HTML would allow that, yes it would be possible security risks. But one option could be an opt-in setting, that the admin have to specify. Embeds and/or iframes would be a good start.

**Comment by arladmin** [comment]
@tommoor 
I found it a bit strange as well: the lack of any capability to embed html/js code.

It's one of the primary ways to extend a tool's native feature-set, instead, it seems the focus here is primarily on pre-defined (static) plugins, which effectively gatekeeps and severely constrains the whole ecosystem of this "open source" tool.

**Reply by arladmin** [reply]
@tommoor

**Reply by tommoor** [reply]
We won't be supporting arbitrary HTML as it's a security risk but you can certainly build embeds pattern matching off existing ones.

**Reply by arladmin** [reply]
@tommoor 

Why?

The world does it, safely. (One popular way being, using iframes with oembed)

Also, the onus of the risk is on the self-hoster/account owner, after all.

**Reply by tommoor** [reply]
What kind of thing are you looking to achieve? Better to focus on the problem rather than a solution

**Reply by arladmin** [reply]
> Better to focus on the problem rather than a solution

Rather weird approach.

<hr>

1. Open capability to embed most HTML codes, including tracking scripts, audio/video embed scripts originating from any tool.
2. Additional capability to add certain HTML entities (with inline CSS?), e.g. custom buttons, among other things.
3. Capability to customise CSS of any document

**Reply by tommoor** [reply]
Gotcha, it sounds more like you're after a website generator in some ways.

Outline is quite opinionated software, think of it more like Google Docs than something like Webflow/Frontpage in terms of the functionality and experience we're looking to provide, as it's also primarily developed as a cloud-hosted product that does come with security restrictions too.

**Reply by arladmin** [reply]
And strictly restricting the integration/extending capabilities to only a select few tools/platforms (on the basis of the opinions of the maintainers), conforms the software with "Open-Source Wiki software" standards?

There's still a world of difference between being a highly restrictive wiki/knowledgebase software and a website generator.

**Comment by routerino** [comment]
> @tommoor
> 
> Why?
> 
> The world does it, safely. (One popular way being, using iframes with oembed)
> 
> Also, the onus of the risk is on the self-hoster/account owner, after all.

I'd suggest you google "cross site scripting vulnerability" before saying 'the world does it safely'.

iframes are functionally sandboxed away from the main application and can't be manipulated within the application itself.

**Reply by arladmin** [reply]
> I'd suggest you google "cross site scripting vulnerability" before saying 'the world does it safely'.

Seems like your 'google search' stopped short of actually trying to find solutions.

One example:
https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#xss-prevention-rules-summary

**Reply by ghost** [reply]
My 2c: If Notion does it safely, Outline can do it. They have chosen not to. Which is up to them. But we certainly won't be using Outline as a result.

**Reply by tommoor** [reply]
Could you point to the functionality in Notion? I wonder if I'm misunderstanding the request

**Comment by ghost** [comment]
> Could you point to the functionality in Notion? I wonder if I'm misunderstanding the request

Its hard to find the feature listed in the Notion docs. I think I found it by mistake. This looks to be an example: https://nickjanes.notion.site/How-to-Create-Custom-Widgets-and-Embed-in-Notion-de75fba83e954a1480281ab5dfed55eb.

Probably best to see for yourself, just create a free Notion account, press the plus on a page in there and choose 'embed' then paste any site's URL and the Notion server will get the HTML from the URL then create a block in your Notion page with the HTML from the URL.

We are going to use it for windows of data showing live site info (release version, date etc).

**Reply by tommoor** [reply]
Yea, we'll support this (It's just blocked by #3000 currently) – it's different to writing arbitrary HTML and JavaScript into the page though which was the original request on this issue.

**Reply by ghost** [reply]
My bad, sorry.

Looking forward to it.
